artiq/test/coredevice/test_moninj.py

```diff
@@ -66,3 +86,175 @@ def injection_status_cb(channel, override, value):
+        
+    @kernel
+    def init_channel(self, channel):
+        self.core.reset()
+        self.core.break_realtime()
+        channel.cpld.init()
+        channel.init()
+
+    @kernel
+    def read_raw(self, channel):
+        self.init_channel(channel)
+        return channel.get_mu()
+
+    @kernel
+    def write_raw(self, channel, ftw: TInt64, pow: TInt32 = 0):
+        self.init_channel(channel)
+        channel.set_mu(ftw, pow)
+
+    @kernel
+    def read(self, channel):
+        self.init_channel(channel)
+        return channel.get()
+
+    @kernel
+    def write(self, channel, freq: TFloat, phase: TFloat = 0.0):
+        self.init_channel(channel)
+        channel.set(freq, phase)
+
+
+class AD991XMonitorTest(ExperimentCase, IsolatedAsyncioTestCase):
+    def get_urukuls(self):
+        urukuls = dict()
+        ddb = self.device_db.get_device_db()
+        for name, desc in ddb.items():
+            if isinstance(desc, dict) and desc["type"] == "local":
+                module, cls = desc["module"], desc["class"]
+                if (module, cls) == ("artiq.coredevice.ad9910", "AD9910"):
+                    urukuls[name] = self.device_mgr.get(name)
+                if (module, cls) == ("artiq.coredevice.ad9912", "AD9912"):
+                    urukuls[name] = self.device_mgr.get(name)
+        return urukuls
+
+    def setUp(self):
+        super().setUp()
+        try:
+            self.core = self.device_mgr.get_desc("core")
+            self.urukuls = self.get_urukuls()
+            self.kernel = self.create(_UrukulExperiment)
+        except KeyError as e:
+            # skip if ddb does not match requirements
+            raise unittest.SkipTest(
+                "test device not available: `{}`".format(*e.args))
+
+    @asynccontextmanager
+    async def open_comm_session(self, notifications=None, core_host=None, auto_monitor=True):
+        if notifications is None:
+            notifications = []
+        if core_host is None:
+            core_host = self.core["arguments"]["host"]
+
+        def monitor_cb(channel, probe, value):
+            notifications.append((channel, probe, value))
+
+        moninj_comm = CommMonInj(monitor_cb, lambda x, y, z: None)
+        try:
+            await moninj_comm.connect(core_host)
+            if auto_monitor:
+                self.setup_monitor(moninj_comm)
+            yield moninj_comm
+        finally:
+            self.setup_monitor(moninj_comm, teardown=True)
+            await moninj_comm._writer.drain()
+            await asyncio.sleep(0.5)
+            await moninj_comm.close()
+
+    def setup_monitor(self, moninj_comm, *, teardown=False, reset_zero=True):
+        for name, dev in self.urukuls.items():
+            moninj_comm.monitor_probe(not teardown, dev.bus.channel, get_last_integers(name))
+            if reset_zero and not teardown:
+                self.kernel.write_raw(dev, 0)
+                ftw, _ = self.kernel.read_raw(dev)
```

Comment:
What is that for?

Reply:
Resets the frequency tuning word to 0, sometimes if it has the same data the monitor probe notifier from firmware would simply not update. This also ensures the probe can be tested in a predictable way with an initial sequence.

Reply:
Sure, the firmware only reports changes. This is correct and consistent behavior, not something that happens "sometimes".
I was asking about ``read_raw`` though.

Reply:
I see...let's get rid of that